### PR TITLE
DYN-7566 Handle python package version mismatch

### PR DESCRIPTION
**Pull request by zeusongit**
### Purpose

The PR handles the case when a graph contains python nodes pointing to a python engine added from a package version which do not match the package version installed in the current session.
- User is prompted once per session, if the graph they are opening has python nodes with such a mismatch.
- An option to update the existing nodes to use the currently installed python engine package version is provided.
- If the user opts to update the nodes, node dependencies are reset and the graph is marked dirty.

When the user choses `No`
DynamoSandbox_DYU8t5MLgK

When the user choses `Yes`
DynamoSandbox_zmsnY1uFnD

### Declarations

Check these if you believe they are true

- [ ] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated
- [ ] This PR contains no files larger than 50 MB

### Release Notes

Handle python package version mismatch in Dynamo

### Reviewers

@DynamoDS/dynamo

### CODE REVIEW

File: src/WorkspaceDependencyViewExtension/WorkspaceDependencyViewExtension.cs [hunk 1]
```diff
@@ -258,6 +286,15 @@ internal void DependencyRegen(WorkspaceModel ws, bool forceCompute = false)
             }
         }
 
+        private Graph.Nodes.NodeModel GetNodeFromGUID(WorkspaceModel ws, Guid guid)
+        {
+            return ws.Nodes.FirstOrDefault(x => x.GUID == guid);
+        }
+        private string GetLoadedPackageVersion(string pkg)
+        {
+            return pmExtension.PackageLoader.LocalPackages.Where(x => x.Name == pkg).FirstOrDefault().VersionName;
```
Comment (twastvedt) [inline src/WorkspaceDependencyViewExtension/WorkspaceDependencyViewExtension.cs:295]: I think you don't need the Where, just put the lambda in FirstOrDefault

File: src/WorkspaceDependencyViewExtension/WorkspaceDependencyViewExtension.cs [hunk 2]
```diff
@@ -229,6 +233,30 @@ internal void DependencyRegen(WorkspaceModel ws, bool forceCompute = false)
                         {
                             packageDependencyInfo.Path = targetInfo.RootDirectory;
                         }
+                        if (!PythonPackageVersionMismatchWarningDismissed)
+                        {
+                            if (packageDependencyInfo.State == PackageDependencyState.IncorrectVersion)
+                            {
+                                //check if a version mismatch for a python node engine has been detected
+                                var pyNodes = packageDependencyInfo.Nodes.Where(x => GetNodeFromGUID(ws,x) is PythonNodeBase);
```
Comment (twastvedt) [inline src/WorkspaceDependencyViewExtension/WorkspaceDependencyViewExtension.cs:241]: I think we can clean this up a bit. We get all the python nodes, then only use the first and end up casting it twice. Maybe:
`packageDependencyInfo.Nodes.Select(x => GetNodeFromGUID(ws, x)).OfType<PythonNodeBase>().FirstOrDefault()`?

**Review by twastvedt — APPROVED** [review]
Just a couple optimizations, otherwise looks good! Do we test dialogs like this? Are there already tests for the underlying functionality?

**Comment by twastvedt** [comment]
Does this dialog need to show up when opening the graph? My concern is that it's a bit opaque - no way to inspect what the Python scripts contain, or to test out how they behave under the installed engine. We don't have this prompt for node package version mismatches...

Could this just be something that happens when the user clicks "keep installed version" in Workspace References? That seems more in keeping with the current pattern for node packages. But happy to discuss!

**Comment by zeusongit** [comment]
> Does this dialog need to show up when opening the graph? My concern is that it's a bit opaque - no way to inspect what the Python scripts contain, or to test out how they behave under the installed engine. We don't have this prompt for node package version mismatches...
> 
> Could this just be something that happens when the user clicks "keep installed version" in Workspace References? That seems more in keeping with the current pattern for node packages. But happy to discuss!

Yes, a part of this update was also to add a separate button in workspace references to update python nodes to use the currently installed version. So users will have an option to update the nodes after dismissing the dialog. We can just have that button and remove the dialog as well, if that makes more sense.

**Comment by twastvedt** [comment]
> a part of this update was also to add a separate button(WIP) in workspace references to update python nodes to use the currently installed version

Is there a use case for updating Python nodes separate from updating/dismissing the package version mismatch warning? Or, conversely, does a user need to be able to dismiss the package version mismatch warning but leave the nodes marked as the original version?

Again, happy to discuss, but it seems to me it could just be that if you click "Keep installed version", the Python nodes are updated to match that version.

**Pull request closed without merging**